### PR TITLE
Update forking interface

### PR DESCRIPTION
**Pull request by war-in**
## Introduced changes

After consulting an interface with @tkumor3 I've changed it to be more adjustable (in the context of cheatcodes).
Cheatcodes can set `ForkStateReader` on the fly which should help in setting up forks in the cairo code

## Checklist

- [ ] Linked relevant issue
- [ ] Updated relevant documentation
- [ ] Added relevant tests
- [ ] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/src/lib.rs [hunk 1]
```diff
@@ -14,13 +14,13 @@ pub mod state;
...
 pub struct CheatnetState {
     cheatcode_state: CheatcodeState,
-    blockifier_state: CachedState<StateReaderProxy>,
+    blockifier_state: CachedState<ExtendedStateReader>,
     pub deploy_salt_base: u32,
 }
 
 impl CheatnetState {
     #[must_use]
-    pub fn new(state: StateReaderProxy) -> Self {
```
Comment (Radinyn) [inline crates/cheatnet/src/lib.rs:23]: I'm wondering whether the `new` method should stick to creating the non-forkable state perhaps creating the `ExtendedStateReader` under the hood but I'm fine with this approach

File: crates/cheatnet/src/forking/state.rs [hunk 2]
```diff
@@ -18,62 +42,88 @@ impl StateReader for ForkStateReader {
...
-            .address_to_nonce
-            .get(&contract_address)
-        {
-            Some(nonce) => Ok(*nonce),
-            None => self.worker.get_nonce(contract_address),
+        match self.runtime.block_on(
+            self.client
+                .get_nonce(self.block_id, contract_address.to_field_element()),
+        ) {
+            Ok(nonce) => Ok(Nonce(StarkFelt::from(nonce))),
```
Comment (Radinyn) [inline crates/cheatnet/src/forking/state.rs:60]: Perhaps nonce should be added to conversions?

File: crates/cheatnet/src/forking/state.rs [hunk 3]
```diff
@@ -18,62 +42,88 @@ impl StateReader for ForkStateReader {
...
-            .class_hash_to_compiled_class_hash
-            .get(&class_hash)
-            .copied()
-            .unwrap_or_default();
-        Ok(compiled_class_hash)
+    fn get_compiled_class_hash(
+        &mut self,
+        _class_hash: ClassHash,
+    ) -> StateResult<CompiledClassHash> {
+        todo!()
```
Comment (Radinyn) [inline crates/cheatnet/src/forking/state.rs:127]: ?

**Review by Radinyn — APPROVED** [review]

**Review by piotmag769 — APPROVED** [review]

File: crates/cheatnet/src/state.rs [hunk 4]
```diff
@@ -67,21 +113,17 @@ impl StateReader for DictStateReader {
...
     ) -> StateResult<StarkFelt> {
         let contract_storage_key = (contract_address, key);
-        let value = self
-            .storage_view
+        self.storage_view
             .get(&contract_storage_key)
             .copied()
-            .unwrap_or_default();
-        Ok(value)
+            .ok_or(StateError::StateReadError(String::new()))
```
[the remark this answers is not in the document]
Comment (war-in) [reply crates/cheatnet/src/state.rs:119]: Because `DictStateReader` was the default option. We return the default in the `ExtendedStateReader` now

**Pull request merged**